app/models/assignment.rb

```diff
@@ -392,25 +392,71 @@ def add_group(new_group_name = nil)
     Grouping.create(group: group, assignment: self)
   end
 
-  def add_group_api(new_group_name = nil)
-    if new_group_name.nil?
-      if group_name_autogenerated
-        group = self.course.groups.new
-        group.save(validate: false)
-        group.group_name = group.get_autogenerated_group_name
-        group.save
-      else
-        raise 'A group name was not provided'
+  def add_group_api(new_group_name = nil, members = [])
+    members ||= []
+    group = nil
+
+    Group.transaction do
+      group = if new_group_name.nil?
+                if members.length == 1 && self.group_max == 1
+                  student_user_name = members.first
+                  if self.is_timed
```

Comment:
This is a great condition to check. Putting it here is a larger divergence from the original logic (which didn't check `is_timed` at all).

So instead of putting it here, add it to the condition on line 401 (`&& !self.is_timed`), so that for timed assessments the previous logic will execute. This also helps remove some of the code duplication.